src/shared/libs/iota/quorum.js

```diff
@@ -0,0 +1,415 @@
+    const { frequency, mostFrequent } = findMostFrequent(validResults);
+
+    return (method, threshold) => {
+        // Include faulty result count while determining percentage
+        const percentage = frequency[mostFrequent] / quorumSize * 100;
+
+        if (percentage > threshold) {
+            return mostFrequent;
+        }
+
+        return fallbackToSafeResult(method);
+    };
+};
+
+/**
+ *   Gets a safe result if quorum result falls below defined threshold
+ *
+ *   @method fallbackToSafeResult
+ *   @param {string} method
+ *
+ *   @returns {boolean}
+ **/
+const fallbackToSafeResult = (method) => {
+    const allowedMethodsMap = {
+        wereAddressesSpentFrom: true,
+        getInclusionStates: false,
+        getBalances: '0',
+        latestSolidSubtangleMilestone: EMPTY_HASH_TRYTES,
+    };
+
+    return allowedMethodsMap[method];
+};
+
+/**
+ *   For a list of nodes, find healthy nodes with size equals to minimum quorum nodes size
+ *
+ *   @method findSyncedNodes
+ *   @param {array} nodes
+ *   @param {array} [syncedNodes = []]
+ *   @param {array} [blacklistedNodes = []]
+ *
+ *   @returns {Promise}
+ **/
+const findSyncedNodes = (nodes, quorumSize, syncedNodes = [], blacklistedNodes = []) => {
+    const sizeOfSyncedNodes = size(syncedNodes);
+    const whitelistedNodes = filter(nodes, (node) => !includes(blacklistedNodes, node) && !includes(syncedNodes, node));
+
+    if (isEmpty(whitelistedNodes)) {
+        return Promise.reject(new Error(Errors.NOT_ENOUGH_HEALTHY_QUORUM_NODES));
+    }
+
+    const selectedNodes =
+        sizeOfSyncedNodes === quorumSize ? syncedNodes : sampleSize(whitelistedNodes, quorumSize - sizeOfSyncedNodes);
+
+    return Promise.all(map(selectedNodes, (provider) => isNodeSynced(provider).catch(() => undefined))).then(
+        (results) => {
+            const { activeNodes, inactiveNodes } = transform(
+                selectedNodes,
+                (acc, node, idx) => (results[idx] ? acc.activeNodes.push(node) : acc.inactiveNodes.push(node)),
+                { activeNodes: [], inactiveNodes: [] },
+            );
+
+            if (size(activeNodes) === size(selectedNodes)) {
+                return [...syncedNodes, ...activeNodes];
+            }
+
+            return findSyncedNodes(
+                nodes,
+                quorumSize,
+                // Add active nodes to synced nodes
+                [...syncedNodes, ...activeNodes],
+                // Add inactive nodes to blacklisted nodes
+                [...blacklistedNodes, ...inactiveNodes],
+            );
+        },
+    );
+};
+
+/**
+ *   From a list of synced nodes, compute a quorum result for wereAddressesSpentFrom iota api
+ *
+ *   @method getQuorumForWereAddressesSpentFrom
+ *   @param {array} payload - Addresses
+ *   @param {array} syncedNodes
+ *
+ *   @returns {Promise}
+ **/
+const getQuorumForWereAddressesSpentFrom = (payload, syncedNodes) => {
```

Comment:
the following 4 (?) functions are duplicates except for the method name (which is used to invoke the API and get the safe default). 

could this be done using the `method` as an argument? and have only 1 copy of this function, the reason being that this mechanism is new and might need to be tweaked, maintaining all the copies is a hassle and would lead to errors.


and later adding more methods to quorum would be easier as well.

Reply:
Great suggestion. 

One reason why I avoided doing that in the first place is because some methods (e.g. getBalances, getNodeInfo) response is not a plain array of booleans like the rest of the methods (wereAddressesSpentFrom, getLatestInclusion). However, it's not a big problem. I agree that all four methods had duplication and maintaining all four would be a hassle. 

Have simplified the implementation. Now, there's only a single [function](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L261) that serves all four endpoints and also a separate [function](https://github.com/laumair/trinity-wallet/blob/feat/quorum/src/shared/libs/iota/quorum.js#L180) for preparing results. 

Reason why I decoupled preparation of results from `#getQuorum` is because it would be a lot easier to unit test both separately.